hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/BaseFileOnlyRelation.scala

```diff
@@ -84,21 +84,24 @@ class BaseFileOnlyRelation(sqlContext: SQLContext,
 
   protected def collectFileSplits(partitionFilters: Seq[Expression], dataFilters: Seq[Expression]): Seq[HoodieBaseFileSplit] = {
     val partitions = listLatestBaseFiles(globPaths, partitionFilters, dataFilters)
-    val fileSplits = partitions.values.toSeq.flatMap { files =>
-      files.flatMap { file =>
-        // TODO move to adapter
-        // TODO fix, currently assuming parquet as underlying format
-        HoodieDataSourceHelper.splitFiles(
-          sparkSession = sparkSession,
-          file = file,
-          // TODO clarify why this is required
-          partitionValues = getPartitionColumnsAsInternalRow(file)
-        )
+    val fileSplits = partitions.values.toSeq
+      .flatMap { files =>
+        files.flatMap { file =>
+          // TODO fix, currently assuming parquet as underlying format
+          HoodieDataSourceHelper.splitFiles(
+            sparkSession = sparkSession,
+            file = file,
+            partitionValues = getPartitionColumnsAsInternalRow(file)
+          )
+        }
       }
-    }
+      // NOTE: It's important to order the splits in the reverse order of their
+      //       size so that we can subsequently bucket them in an efficient manner
+      .sortBy(_.length)(implicitly[Ordering[Long]].reverse)
```

Comment:
I'm wondering, instead of sorting the file splits here, whether we can customize the subsequent bucketing algorithm in the adapter.  Also, should the TODOs be kept, since they are not resolved?

Reply:
I see within the spark adaptor we use next fit  bin packing algo. If thats never going to change, we can leave it here. if not, we can move the sorting within adaptor.getFilePartitions(...)

Reply:
We want to maintain parity w/ Spark behavior which simply does greedy packing (which is at most 2x less efficient than optimal)

@nsivabalan fair call, i agree that `getFilePartitions` seems like more appropriate place for this (and if it would be placed there in Spark itself we wouldn't have to fix it ourselves), but the reason i'm placing it in here is to keep our code (which is essentially copy of Spark's) in sync (at least for now)

Reply:
https://issues.apache.org/jira/browse/HUDI-3901

Reply:
@alexeykudinkin we still need to keep the following TODOS, to make sure we don't lose context?
```
// TODO move to adapter
// TODO clarify why this is required
```